File: src/main/java/whilestd/Leaf.java
package whilestd;

import static whilestd.Utils.nil;

public class Leaf extends BinTree {

    private String symbol;

    public Leaf(String symbol) {
        this.symbol = symbol;
    }

    @Override
    protected int toInt() {
        return 0;
    }

    @Override
    protected boolean toBool() {
        return false;
    }

    @Override
    public String toString() {
        return this.symbol;
    }

    @Override
    protected BinTree hd() {
        return nil();
    }

    @Override
    protected BinTree tl() {
        return nil();
    }

    @Override
    public BinTree clone() {
        return new Leaf(this.symbol);
    }

    @Override
    public boolean equals(Object other) {
        if (other instanceof Leaf) {
            Leaf otherLeaf = (Leaf) other;
            return this.symbol.equals(otherLeaf.symbol);
        }
        return false;
    }

    @Override
    protected void prettyPrint() {
        System.out.print(this.symbol);
    }
    
}

File: whilestd/src/Leaf.cpp
#include "whilestd/Leaf.h"
#include "whilestd/Node.h"
#include <memory>
#include <string>
#include <iostream>

namespace whilestd {

Leaf::Leaf(const std::string& symbol) : symbol(symbol) {}

Leaf::operator int() const {
    return 0;
}

Leaf::operator bool() const {
    return false;
}

Leaf::operator std::string() const {
    return symbol;
}

BinTreePtr Leaf::hd() const {
    return std::make_unique<Node>();
}

BinTreePtr Leaf::tl() const {
    return std::make_unique<Node>();
}

BinTreePtr Leaf::clone() const {
    return std::make_unique<Leaf>(symbol);
}

std::ostream& Leaf::pp(std::ostream& os) const {
    return os << symbol;
}

bool Leaf::equals(const BinTreePtr& other) const {
    if (auto otherLeaf = dynamic_cast<Leaf*>(other.get())) {
        return symbol == otherLeaf->symbol;
    }
    return false;
}

}
